#**Gender Prediction through Voice**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#reading the dataset
df = pd.read_csv('/content/voice.csv')

df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3168 entries, 0 to 3167
Data columns (total 21 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   meanfreq  3168 non-null   float64
 1   sd        3168 non-null   float64
 2   median    3168 non-null   float64
 3   Q25       3168 non-null   float64
 4   Q75       3168 non-null   float64
 5   IQR       3168 non-null   float64
 6   skew      3168 non-null   float64
 7   kurt      3168 non-null   float64
 8   sp.ent    3168 non-null   float64
 9   sfm       3168 non-null   float64
 10  mode      3168 non-null   float64
 11  centroid  3168 non-null   float64
 12  meanfun   3168 non-null   float64
 13  minfun    3168 non-null   float64
 14  maxfun    3168 non-null   float64
 15  meandom   3168 non-null   float64
 16  mindom    3168 non-null   float64
 17  maxdom    3168 non-null   float64
 18  dfrange   3168 non-null   float64
 19  modindx   3168 non-null   float64
 20  label     3168 non-null   obje

In [ ]:
df['label'].value_counts()
#balance class

label
male      1584
female    1584
Name: count, dtype: int64

In [ ]:
x = df.iloc[:,:-1].values
y = df['label'].values

In [ ]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

#normalisation
scaler = StandardScaler()
x = scaler.fit_transform(x)

#encoding
label = LabelEncoder()
y = label.fit_transform(y)

In [ ]:
x[:5] #an array and normalized

array([[-4.04924806,  0.4273553 , -4.22490077, -2.57610164, -5.69360723,
        -0.21477826,  2.29330585,  1.76294635, -0.03908279,  0.4715753 ,
        -2.14121031, -4.04924806, -1.81203825, -1.0979981 ,  0.56595854,
        -1.5642046 , -0.70840431, -1.43142165, -1.41913712, -1.45477229],
       [-3.84105325,  0.6116695 , -3.99929342, -2.48688452, -5.58898726,
        -0.25848536,  4.54805598,  4.43300778, -0.06523603,  0.59443122,
        -2.14121031, -3.84105325, -1.07959443, -1.09153262, -0.29403034,
        -1.56191576, -0.70840431, -1.41810716, -1.4058184 , -1.01410294],
       [-3.46306647,  1.60384791, -4.09585052, -2.7069865 , -3.92869875,
         0.90932569,  6.51365636,  7.32620677, -1.08373029,  0.39826097,
        -2.14121031, -3.46306647, -1.36536788, -1.10039656,  0.41048032,
        -1.56386648, -0.70840431, -1.42920257, -1.41691733, -1.06534356],
       [-0.99215739,  0.89999793, -0.75945371, -0.90141834, -0.7112046 ,
         0.63269035, -0.44985837, -0.24009931,  

In [ ]:
y #labelled

array([1, 1, 1, ..., 0, 0, 0])

In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)
x_train.shape

(2534, 20)

In [ ]:
#modeling
#SVM Kernelling - Linear, Polynomial, RBF(Radial Based Function), Sigmoid
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

SVM_bank = SVC() #default is RBF

#training

SVM_bank.fit(x_train, y_train)

#prediction

y_hat = SVM_bank.predict(x_test)

#evaluation

print('The accuracy based on RBF kernel is:', (accuracy_score(y_test, y_hat))*100)


The accuracy based on RBF kernel is: 98.42271293375394


In [ ]:
#modeling
#SVM Kernelling - Linear, Polynomial, RBF(Radial Based Function), Sigmoid
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

SVM_bank = SVC(kernel = 'linear')

#training

SVM_bank.fit(x_train, y_train)

#prediction

y_hat = SVM_bank.predict(x_test)

#evaluation

print('The accuracy based on Linear kernel is:', (accuracy_score(y_test, y_hat))*100)


The accuracy based on Linear kernel is: 97.3186119873817


In [ ]:
#modeling
#SVM Kernelling - Linear, Polynomial, RBF(Radial Based Function), Sigmoid
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

SVM_bank = SVC(kernel = 'poly')

#training

SVM_bank.fit(x_train, y_train)

#prediction

y_hat = SVM_bank.predict(x_test)

#evaluation

print('The accuracy based on Poly kernel is:', (accuracy_score(y_test, y_hat))*100)


The accuracy based on Poly kernel is: 95.89905362776025


In [ ]:
#modeling
#SVM Kernelling - Linear, Polynomial, RBF(Radial Based Function), Sigmoid
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

SVM_bank = SVC(kernel = 'sigmoid')

#training

SVM_bank.fit(x_train, y_train)

#prediction

y_hat = SVM_bank.predict(x_test)

#evaluation

print('The accuracy based on Sigmoid kernel is:', (accuracy_score(y_test, y_hat))*100)


The accuracy based on Sigmoid kernel is: 79.02208201892745


In [ ]:
#cross validation - mean value of accuracy

from sklearn.model_selection import cross_val_score
import numpy as np

SVM_bank = SVC()

mean_accuracy = cross_val_score (SVM_bank, x, y, cv = 5, scoring = 'accuracy', verbose = 2)
# verbose = for deep learning (1 = only show the process)
# cv = how many times to run the dataset randomly (shuffling randomly)


print('The total number of accuracy are:', mean_accuracy)
print('')
print('The average of all is:', (np.mean(mean_accuracy))*100)

[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
The total number of accuracy are: [0.94006309 0.97003155 0.98107256 0.98894155 0.95734597]

The average of all is: 96.74909424352516


In [ ]:
#cross validation for SVM linear kernel
SVM_bank = SVC(kernel = 'linear')

mean_accuracy = cross_val_score (SVM_bank, x, y, cv = 20, scoring = 'accuracy', verbose = 2)

print('The total number of accuracy are:', mean_accuracy)
print('')
print('The average of all is:', np.mean((mean_accuracy))*100)

[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END ...................

In [ ]:
#cross validation for SVM polynomial kernel
SVM_bank = SVC(kernel = 'poly')

mean_accuracy = cross_val_score (SVM_bank, x, y, cv = 5, scoring = 'accuracy', verbose = 2)

print('The total number of accuracy are:', mean_accuracy)
print('')
print('The average of all is:', (np.mean(mean_accuracy))*100)

[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
The total number of accuracy are: [0.90851735 0.92271293 0.97160883 0.96840442 0.92575039]

The average of all is: 93.93987870089354


In [ ]:
#cross validation for SVM sigmoid kernel
SVM_bank = SVC(kernel = 'sigmoid')

mean_accuracy = cross_val_score (SVM_bank, x, y, cv = 5, scoring = 'accuracy', verbose = 2)
#verbose = for deep learning (1 = only show the process)
# cv = how many times to run the dataset randomly (shuffling randomly)


print('The total number of accuracy are:', mean_accuracy)
print('')
print('The average of all is:', (np.mean(mean_accuracy))*100)

[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.2s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.2s
[CV] END .................................................... total time=   0.2s
The total number of accuracy are: [0.70347003 0.85804416 0.84227129 0.77567141 0.70616114]

The average of all is: 77.71236064805817


###**Prediction of Voice using KNN Algorithm**

In [ ]:
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [ ]:
x1 = df.iloc[:,:-1].values
y1 = df['label'].values

In [ ]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

#normalisation

scaler = StandardScaler()

x1 = scaler.fit_transform(x1)

#encoding

y1 = label.fit_transform(y1)

In [ ]:
#splitting dataset for training and testing
from sklearn.model_selection import train_test_split

x1_train, x1_test, y1_train, y1_test = train_test_split(x1, y1, train_size = 0.8, random_state = 24)

print(x1_train.shape)
print(x1_test.shape)

(2534, 20)
(634, 20)


In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

#training
neigh = KNeighborsClassifier()
neigh.fit(x1_train, y1_train)

#prediction
y1_hat = neigh.predict(x1_test)

#evaluation
print('The accuracy score is:', (accuracy_score(y1_test, y1_hat))*100)


The accuracy score is: 96.84542586750788


In [ ]:
#cross validation for KNN
neigh = KNeighborsClassifier()

mean_accuracy = cross_val_score(neigh, x1, y1, cv = 20, scoring = 'accuracy', verbose = 2)

print('The total accuracy score is:', mean_accuracy)
print('')
print('The average of all is:', (np.mean(mean_accuracy))*100)

[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.2s
[CV] END .................................................... total time=   0.2s
[CV] END .................................................... total time=   0.2s
[CV] END .................................................... total time=   0.2s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END ...................

###**Prediction of Voice using Decision Tree Algorithm**

In [ ]:
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [ ]:
x2 = df.iloc[:,:-1].values
y2 = df['label'].values

In [ ]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

label = LabelEncoder()

y2 = label.fit_transform(y2)

#NORMALISATION

scaler = StandardScaler()

x2 = scaler.fit_transform(x2)

In [ ]:
#splitting dataset for training and testing
from sklearn.model_selection import train_test_split

x2_train, x2_test, y2_train, y2_test = train_test_split(x2, y2, train_size = 0.8, random_state = 24 )

print(x2_train.shape)
print(x2_test.shape)

(2534, 20)
(634, 20)


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

#training
tree = DecisionTreeClassifier()
tree.fit(x2_train, y2_train)

#prediction
y2_hat = tree.predict(x2_test)

#evaluation

print('The accuracy score is:', (accuracy_score(y2_test, y2_hat))*100)


The accuracy score is: 96.84542586750788


In [ ]:
#cross validation for Decision Tree Algorithm

tree = DecisionTreeClassifier()

mean_accuracy = cross_val_score (tree, x2, y2, cv = 5, scoring = 'accuracy', verbose = 2)

print('The total accuracy score is:', mean_accuracy)
print('')
print('The mean accuracy score is:', (np.mean(mean_accuracy))*100)

[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
[CV] END .................................................... total time=   0.1s
The total accuracy score is: [0.90378549 0.95268139 0.96529968 0.9699842  0.93206951]

The mean accuracy score is: 94.47640547988897


###**Prediction of Voice using Logistic Regression Algorithm**

In [ ]:
x3 = df.iloc[:,:-1].values
y3 = df['label'].values

In [ ]:
from sklearn.preprocessing import LabelEncoder, StandardScaler

#labeling target value

label = LabelEncoder()
y3 = label.fit_transform(y3)

#normalisation

scaler = StandardScaler()
x3 = scaler.fit_transform(x3)

In [ ]:
from sklearn.model_selection import train_test_split

x3_train, x3_test, y3_train, y3_test = train_test_split(x3, y3, train_size = 0.8, random_state = 24)

print(x3_train.shape)
print(x3_test.shape)

(2534, 20)
(634, 20)


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

#training
LR = LogisticRegression()
LR.fit(x3_train, y3_train)

#prediction
y3_hat = LR.predict(x3_test)

#evaluation
print('The accuracy score is:', (accuracy_score(y3_test, y3_hat))*100)


The accuracy score is: 96.84542586750788


In [ ]:
#cross validation

LR = LogisticRegression()

mean_accuracy = cross_val_score(LR, x3, y3, cv = 5, scoring = 'accuracy', verbose = 2)

print('The total accuracy score is:', mean_accuracy)
print('The mean of all accuracy is:', (np.mean(mean_accuracy))*100)

[CV] END .................................................... total time=   0.0s
[CV] END .................................................... total time=   0.0s
[CV] END .................................................... total time=   0.0s
[CV] END .................................................... total time=   0.0s
[CV] END .................................................... total time=   0.1s
The total accuracy score is: [0.93217666 0.97476341 0.97318612 0.98104265 0.97156398]
The mean of all accuracy is: 96.6546563607278



**1) The Accuracy Score:-**

a) SVM:

	I) RBF = 98.42
	II) Linear = 97.31
	III) Poly = 95.89
	IV) Sigmoid = 79.02

b) KNN: 96.84

c) Decision Tree: 96.84

d) Logistic Regression: 96.84

____________________________________________

2) The mean accuracy score (CV = 5):-

a) SVM:

	I) RBF = 96.74
	II) Linear = 96.87
	III) Poly = 93.93
	IV) Sigmoid = 77.71

b) KNN: 95.52

c) Decision Tree: 94.47

d) Logistic Regression: 96.65

In conclusion, eventhough the RBF model predicts the most accurate at first compared to other algorithms, the accuracy of linear prediction surpasses RBF and holds the record of most accurate prediction after cross validation.